pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -260,6 +261,10 @@ def save_checkpoint(self, trainer, unused: Optional = None):
         # Mode 3: save last checkpoints
         self._save_last_checkpoint(trainer, monitor_candidates)
 
+        # notify loggers
+        if trainer.is_global_zero and trainer.logger and hasattr(trainer.logger, 'after_save_checkpoint'):
+            trainer.logger.after_save_checkpoint(proxy(self))
```

Comment:
The problem there is that the checkpoint has not been saved yet. We just gather states.
’after_save_checkpoint’ needs to happen after the model has been saved locally.

Reply:
@awaelchli Here is the stack when saving a checkpoint. Items at the same level are called sequentially.

* `ModelCheckpoint.save_checkpoint` → `_save_top_k_checkpoint` and `_save_none_monitor_checkpoint` and `_save_last_checkpoint` → `ModelCheckpoint._save_model` → `ModelCheckpoint._do_save` → `Trainer.save_checkpoint`
  * → `CheckpointConnector.save_checkpoint`
    * → `CheckpointConnect.dump_checkpoint` only gather states through `on_save_checkpoint`
    * → `atomic_save` saves the file

So basically if you prefer it in the checkpoint connector it could be at the end of `CheckpointConnector.save_checkpoint` (after `atomic_save`).

The issue is that it won't be as easy to get access to `ModelCheckpoint` from `logger.after_save_checkpoint` (which we need especially to access checkpoint path):
* calling `after_save_checkpoint` from `ModelCheckpoint` let us pass directly the `ModelCheckpoint` object
* calling `after_save_checkpoint` from `CheckpointConnector` requires a more complex logic, probably also connecting the trainer to the logger with `logger.connect` as we had previously

Let me know how you want me to proceed.

Reply:
I definitely think that it should not be the responsibility of the model checkpoint callback to manage this logging logic.
Instead, I propose to let the trainer/connector orchestrate this. How about this:

In ModelCheckpoint:

```python
def after_save_checkpoint(self, trainer):
     # here you can access trainer.checkpoint_callback

```

In CheckpointConnector:

```python
    def save_checkpoint(self, filepath, weights_only: bool = False):
        ...
        atomic_save(checkpoint, filepath)
        # after save:
        self.trainer.logger.after_save_checkpoint(self.trainer)
```

Reply:
I've been trying to move the logic to `CheckpointConnector.save_checkpoint` but ran into an issue with this approach.

The parameters from `ModelCheckpoint` ("best_model_path", "last_model_path", etc) are updated only later, at [these lines](https://github.com/PyTorchLightning/pytorch-lightning/blob/6a1414681107cf77d1ebeba86359df7b6668bf49/pytorch_lightning/callbacks/model_checkpoint.py#L280-L284).

Those are required to know if the checkpoint should be saved or not, and if the current score improved (to define which is the "best" checkpoint) so it needs to happen after (or at the end of) `ModelCheckpoint.save_checkpoint`.

Reply:
sorry, didn't think of that. I will help brainstorming and fixing the issue after iccv deadline on thursday. Sorry that this is blocking you

Reply:
No worry, I definitely want it to be implemented the best way so let me know if you want me to try another approach!

Reply:
I played around with this. I have an idea how to resolve it, but for this we would need to remember which checkpoints we have already uploaded and which not. That should be possible, right?

Reply:
Yes, I use `self._logged_model_time` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/fda377f6c63417ee8a8c6e4a519b8912d3e37044/pytorch_lightning/loggers/wandb.py#L157) for this purpose.

It captures the path as a key and timestamp as value (as we could have sometimes same file name with newer chekpoint).

Then the logic to find the new checkpoints to log is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/fda377f6c63417ee8a8c6e4a519b8912d3e37044/pytorch_lightning/loggers/wandb.py#L250-L258).

It may happen after we saved a checkpoint or only at the end of training (`log_model=True`) where we may have several checkpoints to log (for example with `save_top_k=3`).

Reply:
would the call to `wandb.log_artifact` upload the files asynchronoulsly?

Reply:
Yes!

Reply:
Just a note related to using `Callback`, we would still need a new hook as it needs to happen after saving the checkpoint (so `on_save_checkpoint` would not be the right place).

Reply:
Just checking for an update @awaelchli . I think many people are interested in this feature :smiley: 

Right now this current version is the cleanest way I found. Would it make sense to merge it and update later if there is some other functionality (logger considered as regular callbacks, new "after_save" method for callback, etc)?

Reply:
Sorry for letting you hang for a few days on this! Yes it is indeed a nice feature and I myself am a heavy wandb user. So certainly appreciate it.
I do not see a more elegant solution right now without larger changes, but I don't want to block it until we find one. 
@PyTorchLightning/core-contributors let us know what you think. In summary, the question is:
Is it acceptable that the `ModelCheckpoint` reaches through the Trainer into the logger to trigger a notification when saving is complete, so that the logger can upload the checkpoint files?

Reply:
If yes, I think we should go ahead with this PR. Uploading the model to the WandB cloud and keeping it versioned together with the logs is a very nice feature to have from the perspective of reproducibility

Reply:
Let me know if we should proceed and I can merge master to keep this branch in sync

Reply:
I will ping the team internally